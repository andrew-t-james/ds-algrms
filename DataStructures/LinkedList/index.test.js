const LinkedList = require('.')

describe('LinkedList', () => {
  test('should append new Node into LinkedList', () => {
    const list = new LinkedList()
    expect(list.head).toBe(null)
    list.append(1)
    expect(list.head.data).toBe(1)
    list.append(2)
    expect(list.head.next.data).toBe(2)
  })

  test('should append Node at a position in LinkedList', () => {
    const list = new LinkedList()
    const expected = {
      head: {
        data: 1,
        next: { data: 3, next: { data: 2, next: null } }
      }
    }
    list.append(1)
    list.append(2)
    list.appendAt(1, 3)
    expect(list).toEqual(expected)
  })

  test('should return length of LinkedList', () => {
    const list = new LinkedList()
    list.append(1)
    expect(list.length()).toBe(1)
    list.append(2)
    expect(list.length()).toBe(2)
  })

  test('should return boolean based on LinkedList length', () => {
    const list = new LinkedList()
    expect(list.empty()).toBe(true)
    list.append(1)
    expect(list.empty()).toBe(false)
  })

  test('should return display Nodes and return sting of data from Nodes', () => {
    const list = new LinkedList()
    expect(list.empty()).toBe(true)
    list.append(1)
    list.append(2)
    list.append(3)
    expect(list.display()).toBe('1 2 3')
  })

  test('should return traverse Nodes and update data of that Node', () => {
    const list = new LinkedList()
    const expected = {
      head: {
        data: 11,
        next: { data: 12, next: { data: 13, next: null } }
      }
    }
    list.traverse(node => {
      node.data += 10000
    })
    expect(list).toEqual({ head: null })
    list.append(1)
    list.append(2)
    list.append(3)
    list.traverse(node => {
      node.data += 10
    })
    expect(list).toEqual(expected)
  })

  test('should return false || index if value exists in list', () => {
    const list = new LinkedList()
    expect(list.empty()).toBe(true)
    list.append(2)
    list.append(1)
    expect(list.search(1)).toBe(1)
  })

  test('should return remove data from head of LinkedList', () => {
    let list = new LinkedList()
    const expected = {
      head: {
        data: 2,
        next: {
          data: 3,
          next: null
        }
      }
    }
    expect(list.empty()).toBe(true)
    list.append(1)
    list.append(2)
    list.append(3)
    list.remove(1)
    expect(list.length()).toBe(2)
    expect(list).toEqual(expected)
  })

  test('should return remove data from tail of LinkedList', () => {
    let list = new LinkedList()
    const expected = {
      head: {
        data: 1,
        next: {
          data: 2,
          next: null
        }
      }
    }
    expect(list.empty()).toBe(true)
    list.append(1)
    list.append(2)
    list.append(3)
    list.remove(3)
    expect(list.length()).toBe(2)
    expect(list).toEqual(expected)
  })

  test('should return remove data from LinkedList', () => {
    let list = new LinkedList()
    const expected = {
      head: {
        data: 1,
        next: {
          data: 3,
          next: null
        }
      }
    }
    expect(list.empty()).toBe(true)
    list.append(1)
    list.append(2)
    list.append(3)
    list.remove(2)
    expect(list.length()).toBe(2)
    expect(list).toEqual(expected)
  })
})
