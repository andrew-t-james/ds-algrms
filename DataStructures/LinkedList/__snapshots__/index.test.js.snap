// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkedList should delete node by value 1`] = `
LinkedList {
  "head": Node {
    "data": 1,
    "next": Node {
      "data": 2,
      "next": Node {
        "data": 3,
        "next": null,
      },
    },
  },
}
`;

exports[`LinkedList should delete node by value 2`] = `
LinkedList {
  "head": Node {
    "data": 2,
    "next": Node {
      "data": 3,
      "next": null,
    },
  },
}
`;

exports[`LinkedList should delete node by value 3`] = `
LinkedList {
  "head": Node {
    "data": 1,
    "next": Node {
      "data": 2,
      "next": Node {
        "data": 3,
        "next": null,
      },
    },
  },
}
`;

exports[`LinkedList should delete node by value 4`] = `
LinkedList {
  "head": Node {
    "data": 1,
    "next": Node {
      "data": 3,
      "next": null,
    },
  },
}
`;

exports[`LinkedList should detect a circular reference in a linked list 1`] = `
LinkedList {
  "head": Node {
    "data": 1,
    "next": Node {
      "data": 2,
      "next": Node {
        "data": 3,
        "next": null,
      },
    },
  },
}
`;

exports[`LinkedList should remove duplicate nodes in a linked list 1`] = `
LinkedList {
  "head": Node {
    "data": 1,
    "next": Node {
      "data": 2,
      "next": Node {
        "data": 2,
        "next": Node {
          "data": 3,
          "next": Node {
            "data": 3,
            "next": Node {
              "data": 4,
              "next": Node {
                "data": 4,
                "next": null,
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`LinkedList should remove duplicate nodes in a linked list 2`] = `
LinkedList {
  "head": Node {
    "data": 1,
    "next": Node {
      "data": 2,
      "next": Node {
        "data": 3,
        "next": Node {
          "data": 4,
          "next": null,
        },
      },
    },
  },
}
`;

exports[`LinkedList should return a union of two liked lists 1`] = `
LinkedList {
  "head": Node {
    "data": 1,
    "next": Node {
      "data": 2,
      "next": Node {
        "data": 3,
        "next": null,
      },
    },
  },
}
`;

exports[`LinkedList should return a union of two liked lists 2`] = `
LinkedList {
  "head": Node {
    "data": 1,
    "next": Node {
      "data": 2,
      "next": Node {
        "data": 3,
        "next": null,
      },
    },
  },
}
`;

exports[`LinkedList should return a union of two liked lists 3`] = `
LinkedList {
  "head": Node {
    "data": 4,
    "next": Node {
      "data": 5,
      "next": Node {
        "data": 6,
        "next": null,
      },
    },
  },
}
`;

exports[`LinkedList should return a union of two liked lists 4`] = `
LinkedList {
  "head": Node {
    "data": 1,
    "next": Node {
      "data": 2,
      "next": Node {
        "data": 3,
        "next": Node {
          "data": 4,
          "next": Node {
            "data": 5,
            "next": Node {
              "data": 6,
              "next": null,
            },
          },
        },
      },
    },
  },
}
`;

exports[`LinkedList should return mid node 1`] = `
LinkedList {
  "head": Node {
    "data": 1,
    "next": Node {
      "data": 2,
      "next": Node {
        "data": 3,
        "next": Node {
          "data": 4,
          "next": null,
        },
      },
    },
  },
}
`;

exports[`LinkedList should return mid node 2`] = `
LinkedList {
  "head": Node {
    "data": 1,
    "next": Node {
      "data": 2,
      "next": Node {
        "data": 3,
        "next": Node {
          "data": 4,
          "next": Node {
            "data": 5,
            "next": null,
          },
        },
      },
    },
  },
}
`;

exports[`LinkedList should return node that is an intersection in multiple liked lists 1`] = `
LinkedList {
  "head": Node {
    "data": 1,
    "next": Node {
      "data": 2,
      "next": Node {
        "data": 3,
        "next": null,
      },
    },
  },
}
`;

exports[`LinkedList should return node that is an intersection in multiple liked lists 2`] = `
LinkedList {
  "head": Node {
    "data": 3,
    "next": Node {
      "data": 4,
      "next": Node {
        "data": 5,
        "next": null,
      },
    },
  },
}
`;

exports[`LinkedList should return node that is an intersection in multiple liked lists 3`] = `
LinkedList {
  "head": Node {
    "data": 3,
    "next": null,
  },
}
`;

exports[`LinkedList should return nth node from the last node 1`] = `
LinkedList {
  "head": Node {
    "data": 1,
    "next": Node {
      "data": 2,
      "next": Node {
        "data": 3,
        "next": Node {
          "data": 4,
          "next": Node {
            "data": 5,
            "next": Node {
              "data": 6,
              "next": null,
            },
          },
        },
      },
    },
  },
}
`;

exports[`LinkedList should reverse linked list 1`] = `
LinkedList {
  "head": Node {
    "data": 1,
    "next": Node {
      "data": 2,
      "next": Node {
        "data": 3,
        "next": null,
      },
    },
  },
}
`;

exports[`LinkedList should reverse linked list 2`] = `
LinkedList {
  "head": Node {
    "data": 3,
    "next": Node {
      "data": 2,
      "next": Node {
        "data": 1,
        "next": null,
      },
    },
  },
}
`;
